typescript/infra/scripts/warp-routes/monitor-warp-routes-balances.ts

```diff
@@ -0,0 +1,131 @@
+import { Connection } from '@solana/web3.js';
+import { ethers } from 'ethers';
+import { Gauge, Registry } from 'prom-client';
+
+import {
+  ERC20__factory,
+  SealevelHypCollateralAdapter,
+} from '@hyperlane-xyz/hyperlane-token';
+import { ChainMap, MultiProvider } from '@hyperlane-xyz/sdk';
+import { debug, objMap, promiseObjAll } from '@hyperlane-xyz/utils';
+
+import { startMetricsServer } from '../../src/utils/metrics';
+
+import { WarpTokenConfig, tokenList } from './token_config';
+
+const metricsRegister = new Registry();
+const warpRouteTokenBalance = new Gauge({
+  name: 'hyperlane_warp_route_token_balance',
+  help: 'HypERC20 token balance of a Warp Route',
+  registers: [metricsRegister],
+  labelNames: [
+    'chain_name',
+    'token_address',
+    'token_name',
+    'wallet_address',
+    'token_type',
+  ],
+});
+
+async function main(): Promise<boolean> {
+  startMetricsServer(metricsRegister);
+
+  const checkFreqeuncy = 1000; // 1 second
+
+  const multiProvider = new MultiProvider();
+
+  setInterval(async () => {
+    console.log('Checking balances');
+    const balances = await checkBalance(tokenList, multiProvider);
+    await updateTokenBalanceMetrics(tokenList, balances);
+  }, checkFreqeuncy);
+  return true;
+}
+
+async function checkBalance(
+  tokenConfig: WarpTokenConfig,
+  multiprovider: MultiProvider,
+): Promise<ChainMap<number>> {
+  const output: ChainMap<Promise<number>> = objMap(
+    tokenConfig,
+    async (chain, token) => {
+      const provider = multiprovider.getProvider(chain);
+      if (token.type === 'native') {
+        if (token.protocolType === 'ethereum') {
+          const nativeBalance = await provider.getBalance(
+            token.hypNativeAddress,
+          );
+          return parseFloat(
+            ethers.utils.formatUnits(nativeBalance, token.decimals),
+          );
+        } else {
+          // TODO - solana native
+          return 0;
```

Comment:
Should we be using the adapters here? or better yet, a MultiProtocolApp like I've done for HelloWorld here: https://github.com/hyperlane-xyz/hyperlane-monorepo/blob/rossy/multi-protocol-kathy/typescript/helloworld/src/multiProtocolApp/multiProtocolApp.ts#L16

We would need to merge my kathy branch first but we should, it's overdue

Reply:
https://github.com/hyperlane-xyz/hyperlane-monorepo/issues/2708